internal/docstore/dynamodocstore/dynamo.go

```diff
@@ -61,45 +62,159 @@ func (c *collection) RunActions(ctx context.Context, actions []*driver.Action, u
 	if unordered {
 		panic("unordered unimplemented")
 	}
+	groups := c.splitActions(actions)
+	nRun := 0 // number of actions successfully run
+	var err error
+	for _, g := range groups {
+		if g[0].Kind == driver.Get {
+			err = c.runGets(ctx, g)
+		} else {
+			err = c.runWrites(ctx, g)
+		}
+		if err != nil {
+			return driver.ActionListError{{nRun, err}}
+		}
+		nRun += len(g)
+	}
+	return nil
+}
+
+// splitActions divides the actions slice into sub-slices, each of which can be
+// passed to run a dynamo transaction operation.
+// splitActions doesn't change the order of the input slice.
+func (c *collection) splitActions(actions []*driver.Action) [][]*driver.Action {
+	var (
+		groups [][]*driver.Action      // the actions, split; the return value
+		cur    []*driver.Action        // the group currently being constructed
+		wm     = make(map[string]bool) // writes group cannot contain duplicate items
+	)
+	collect := func() { // called when the current group is known to be finished
+		if len(cur) > 0 {
+			groups = append(groups, cur)
+			cur = nil
+			wm = make(map[string]bool)
+		}
+	}
+	for _, a := range actions {
+		if len(cur) > 0 && c.shouldSplit(cur[len(cur)-1], a, wm) ||
+			len(cur) >= 10 { // each transaction can run up to 10 operations.
+			collect()
+		}
+		cur = append(cur, a)
+		if a.Kind != driver.Get {
+			wm[c.primaryKey(a)] = true
+		}
+	}
+	collect()
+	return groups
+}
+
+func (c *collection) shouldSplit(curr, next *driver.Action, wm map[string]bool) bool {
+	if (curr.Kind == driver.Get) != (next.Kind == driver.Get) { // different kind
+		return true
+	}
+	if curr.Kind == driver.Get { // both are Get's
+		return false
+	}
+	_, ok := wm[c.primaryKey(next)]
+	return ok // different Write's in one transaction cannot target the same item
+}
+
+// primaryKey tries to get the primary key from the doc, which is the partition
+// key if there is no sort key, or the combination of both keys. If there is not
+// a key for Create action, it generates a partition key of the doc. It returns
+// the composite key to be guaranteed unique.
+func (c *collection) primaryKey(a *driver.Action) string {
+	pkey, err := a.Doc.GetField(c.partitionKey)
+	if err != nil {
+		if c.ErrorCode(err) == gcerr.NotFound && a.Kind == driver.Create {
+			newPartitionKey := driver.UniqueString()
+			a.Doc.SetField(c.partitionKey, newPartitionKey)
+			return newPartitionKey
+		}
+		panic(err) // shouldn't happen
+	}
+	if c.sortKey != "" {
+		skey, err := a.Doc.GetField(c.sortKey)
+		if err != nil {
+			panic(err) // shouldn't happen
+		}
+		return fmt.Sprintf("%s, %s", pkey, skey)
```

Comment:
This could give the same string for two different items: `a, b, c` could be pkey=`a, b` and skey=`c`, or pkey=`a` and skey=`b, c`.

Also, the keys need not be strings.

I think this function should return a `[2]interface{}`.